### PR TITLE
Introduce the concept of asynchronicity to the service extensions.

### PR DESCRIPTION
**Pull request by Hixie**
This allows us, for example, to wait for the slow mode banner to have
been removed from the screen before triggering a screen shot.

### CODE REVIEW

File: packages/flutter/lib/src/foundation/print.dart [hunk 1]
```diff
@@ -71,11 +72,19 @@ void _debugPrintTask() {
...
     _debugPrintStopwatch.start();
+    _debugPrintCompleter?.complete();
+    _debugPrintCompleter = null;
   }
 }
 
+/// A Future that resolves when there is no longer any buffered content being
+/// printed by [debugPrintThrottled] (which is the default implementation for
+/// [debugPrint], which is used to report errors to the console).
+Future<Null> get debugPrintDone => _debugPrintCompleter.future ?? new Future<Null>.value();
```
Comment (johnmccutchan) [inline packages/flutter/lib/src/foundation/print.dart]: Did you mean: `_debugPrintCompleter?.future` ?

File: packages/flutter_tools/lib/src/vmservice.dart [hunk 2]
```diff
@@ -567,27 +567,39 @@ class VM extends ServiceObjectOwner {
...
   }
 
-  /// Invoke the RPC and return a ServiceObject response.
-  Future<ServiceObject> invokeRpc(
-      String method, { Map<String, dynamic> params, Duration timeout }) async {
-    Map<String, dynamic> response = await invokeRpcRaw(method, params, timeout);
+  /// Invoke the RPC and return a [ServiceObject] response.
+  Future<ServiceObject> invokeRpc(String method, {
+    Map<String, dynamic> params: const <String, dynamic>{},
+    Duration timeout,
```
Comment (johnmccutchan) [inline packages/flutter_tools/lib/src/vmservice.dart]: do you want to let people pass in timeoutFatal here so it cna be forwarded to invokeRpcRaw?

File: packages/flutter_tools/lib/src/vmservice.dart [hunk 3]
```diff
@@ -750,21 +760,18 @@ class Isolate extends ServiceObjectOwner {
...
-    return vm.invokeRpcRaw(method, params, timeout);
+    return vm.invokeRpcRaw(method, params: params, timeout: timeout, timeoutFatal: timeoutFatal);
   }
 
   /// Invoke the RPC and return a ServiceObject response.
-  Future<ServiceObject> invokeRpc(
-      String method, Map<String, dynamic> params) async {
-    Map<String, dynamic> response = await invokeRpcRaw(method, params);
-    return getFromMap(response);
+  Future<ServiceObject> invokeRpc(String method, Map<String, dynamic> params) async {
```
Comment (johnmccutchan) [inline packages/flutter_tools/lib/src/vmservice.dart]: same about timeoutFatal

**Review by johnmccutchan — APPROVED** [review]
LGTM with some small comments

File: packages/flutter/lib/src/foundation/print.dart (hunk 1 above)
Comment (Hixie) [inline packages/flutter/lib/src/foundation/print.dart]: oops, good catch

File: packages/flutter_tools/lib/src/vmservice.dart (hunk 2 above)
Comment (Hixie) [inline packages/flutter_tools/lib/src/vmservice.dart]: No, because I don't know what to return if it timed out but is not fatal.

**Comment by Hixie** [comment]
Thanks. Fixed the first one, the other two I have left as is per comments above. Will land when Travis goes green unless I hear otherwise.

**Pull request merged**